### ISSUE TITLE
Prioritized pods (ubernetes)

### ISSUE THREAD

**Issue by irfanurrehman** [kubernetes/federation#40]
<a href="https://github.com/juanjoperl"><img src="https://avatars2.githubusercontent.com/u/17222177?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [juanjoperl](https://github.com/juanjoperl)**
_Monday Feb 22, 2016 at 20:35 GMT_
_Originally opened as https://github.com/kubernetes/kubernetes/issues/21705_

----

For a possible workaround to the problem of federation, can kubernetes have this behabour?: the ability the services have several sets of labels to select pods. The first set would be the priority, the second come into play if the priority pods do not respond, and so on.
With this capability, could have pods tagged area, and the services have sorted priority areas

**Comment by irfanurrehman** [comment]
<a href="https://github.com/thockin"><img src="https://avatars0.githubusercontent.com/u/5595220?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [thockin](https://github.com/thockin)**
_Tuesday Feb 23, 2016 at 16:47 GMT_

----

That implies support for this sort of priority at the load-balancers, which I don't think is a broad-enough feature to be a core part of the system.  Or maybe I don't understand the suggestion?

**Comment by irfanurrehman** [comment]
<a href="https://github.com/juanjoperl"><img src="https://avatars2.githubusercontent.com/u/17222177?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [juanjoperl](https://github.com/juanjoperl)**
_Tuesday Feb 23, 2016 at 19:20 GMT_

----

Forgiveness because I am newcomer to kubernetes. My question arises from some days when I read the documentation.

My need is to have two centers, with a few tens of milliseconds of latency between them, and I would like to have a single cluster. I know the future existence of Ubernetes but I was wondering if a partial solution to the current kubernetes be possible. I think my choice would focus on Kube-proxy, so it does not round-robin is applied, but the proxy is driven by labels: round robin over label priority pods, and go to another labeled pods if there is a fault.

nodeA1 --> label_Azone
nodeA2 --> label_Azone
nodeB1---> label_Bzone
nodeB2---> label_Bzone

rc of podX_A --> label_Azone, select node label_Azone   (X=redis by example)
rc of podX_B --> label_Bzone, select node label_Bzone

serviceX_A  --> select pod [label_Azone,label_Bzone]
serviceX_B  --> select pod [label_Bzone,label_Azone]

Not knowing in depth Kubernetes not know if you meant external load balancers or type of service called "LoadBalancer", or perhaps both are the same. I think than modify the kube-proxy all is ok, is this right?

Another option (I´m thinking, sorry) could be that the rolling is performed taking into account the latency, that is, the number of requests to a pod is inversely proportional to the ratio of latencies that are from one node to the set of pods. It would be transparent to users (only set service.spec.sessionAffinity to "latency").

Thank you for this project, which I found great.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/thockin"><img src="https://avatars0.githubusercontent.com/u/5595220?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [thockin](https://github.com/thockin)**
_Wednesday Feb 24, 2016 at 02:50 GMT_

----

> My need is to have two centers, with a few tens of milliseconds of latency between them, and I would like to have a single cluster. I know the future existence of Ubernetes but I was wondering if a partial solution to the current kubernetes be possible. I think my choice would focus on Kube-proxy, so it does not round-robin is applied, but the proxy is driven by labels: round robin over label priority pods, and go to another labeled pods if there is a fault.
> 
> nodeA1 --> label_Azone
> 
> nodeA2 --> label_Azone
> 
> nodeB1---> label_Bzone
> nodeB2---> label_Bzone
> 
> rc of podX_A --> label_Azone, select node label_Azone (X=redis by example)
> rc of podX_B --> label_Bzone, select node label_Bzone
> 
> serviceX_A --> select pod [label_Azone,label_Bzone]
> serviceX_B --> select pod [label_Bzone,label_Azone]

Got it.  We do not have that capability in kube-proxy right now,
though as Ubernetes Lite rolls out, it might start to matter
(@quinton-hoole @justinsb).  The kube-proxy (as of 1.2) is pretty dumb
and doesn't really know whether a backend is "up" or not.  It's just a
VIP.  We _could_ start to bias traffic to zone-local backends, but
we'd have to talk about the heuristic to use.

In other words, traffic that hits a node in Zone A would always choose
a backend in zone A unless no backends are Ready, in which case it
would choose a backend in zone B.  Or maybe s/always/usually/ ?

> Another option (I´m thinking, sorry) could be that the rolling is performed taking into account the latency, that is, the number of requests to a pod is inversely proportional to the ratio of latencies that are from one node to the set of pods. It would be transparent to users (only set service.spec.sessionAffinity to "latency").
> 
> Thank you for this project, which I found great.

We don't really have latency numbers.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/quinton-hoole"><img src="https://avatars0.githubusercontent.com/u/10390785?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [quinton-hoole](https://github.com/quinton-hoole)**
_Wednesday Feb 24, 2016 at 18:30 GMT_

----

@thockin @juanjoperl See also docs/design/federated-services.md in #19313 (sorry, it's not merged yet).  The solution proposed there involves using DNS and multiple A records with client failover rather than trying to magic away the distinction between different latency zones. Feel free to comment on the above PR if you have better ideas or disagree with the approach proposed there.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/quinton-hoole"><img src="https://avatars0.githubusercontent.com/u/10390785?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [quinton-hoole](https://github.com/quinton-hoole)**
_Wednesday Feb 24, 2016 at 18:37 GMT_

----

To be clear, by "magic away" I was referring to having a single IP address for all clients, that somehow dynamically redirects to the closest, non-overloaded, fully functional pod (by some definition of all of those terms) for every possible client.  Building that complexity into kube-proxy is, in my opinion, prone to undesirable mass failures.  I agree with Tim's implied desire to keep kube-proxy as simple as possible to reduce the probability of really bad failure scenarios.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/juanjoperl"><img src="https://avatars2.githubusercontent.com/u/17222177?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [juanjoperl](https://github.com/juanjoperl)**
_Wednesday Feb 24, 2016 at 21:40 GMT_

----

@thockin @quinton-hoole I read the information you said. As expected it is a plan that makes you want to have it already in place.

Regarding load metrics that are mentioned in the document for load balancing, it may not be so necessary to have pods with the ability to auto-scaling and Auto reset. I guess it would be for lower loads than loads that trigger these actions, is it ?.Is in this point kubernetes similar as ubernetes?.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/quinton-hoole"><img src="https://avatars0.githubusercontent.com/u/10390785?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [quinton-hoole](https://github.com/quinton-hoole)**
_Wednesday Feb 24, 2016 at 21:58 GMT_

----

@juanjoperl We have no active work under way yet to add proximity or load-based balancing to Kubernetes/kube-proxy, but if anyone wants to add it, I would suggest that it be consistent with what's described in the above doc for Ubernetes.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/thockin"><img src="https://avatars0.githubusercontent.com/u/5595220?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [thockin](https://github.com/thockin)**
_Thursday Feb 25, 2016 at 21:48 GMT_

----

We could make this a more dynamic thing where the default is "equal balance
across all pods, ignoring load or zone" but a non-realtime process could
get stats (somehow, out of scope) and feed us clues about weights.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/quinton-hoole"><img src="https://avatars0.githubusercontent.com/u/10390785?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [quinton-hoole](https://github.com/quinton-hoole)**
_Thursday Feb 25, 2016 at 22:00 GMT_

----

@thockin yes, that's covered in the above doc, BTW.

**Comment by irfanurrehman** [comment]
cc @juanjoperl
